api/prometheus/v1/remote/remote_api.go

```diff
@@ -0,0 +1,376 @@
+				// TODO(bwplotka): Allow users to disable this check or provide their stats for us to know if it's empty.
+				return accumulatedStats, fmt.Errorf("sent v2 request; "+
+					"got 2xx, but PRW 2.0 response header statistics indicate %v samples, %v histograms "+
+					"and %v exemplars were accepted; assumining failure e.g. the target only supports "+
+					"PRW 1.0 prometheus.WriteRequest, but does not check the Content-Type header correctly",
+					accumulatedStats.Samples, accumulatedStats.Histograms, accumulatedStats.Exemplars,
+				)
+			}
+			// Success!
+			// TODO(bwplotka): Debug log with retry summary?
+			return accumulatedStats, nil
+		}
+
+		var retryableErr retryableError
+		if !errors.As(err, &retryableErr) {
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return accumulatedStats, err
+		}
+
+		if !b.Ongoing() {
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return accumulatedStats, err
+		}
+
+		backoffDelay := b.NextDelay() + retryableErr.RetryAfter()
+		r.opts.logger.Error("failed to send remote write request; retrying after backoff", "err", err, "backoff", backoffDelay)
+		select {
+		case <-ctx.Done():
+			// TODO(bwplotka): More context in the error e.g. about retries.
+			return WriteResponseStats{}, ctx.Err()
+		case <-time.After(backoffDelay):
+			// Retry.
+		}
+	}
+}
+
+func compressPayload(tmpbuf *[]byte, enc Compression, inp []byte) (compressed []byte, _ error) {
+	switch enc {
+	case SnappyBlockCompression:
+		compressed = snappy.Encode(*tmpbuf, inp)
+		if n := snappy.MaxEncodedLen(len(inp)); n > len(*tmpbuf) {
+			// grow the buffer for the next time.
+			*tmpbuf = make([]byte, n)
+		}
+		return compressed, nil
+	default:
+		return compressed, fmt.Errorf("unknown compression scheme [%v]", enc)
+	}
+}
+
+func (r *API) attemptWrite(ctx context.Context, compr Compression, proto WriteProtoFullName, payload []byte, attempt int) (WriteResponseStats, error) {
+	u := r.client.URL("api/v1/write", nil)
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(payload))
+	if err != nil {
+		// Errors from NewRequest are from unparsable URLs, so are not
+		// recoverable.
+		return WriteResponseStats{}, err
+	}
+
+	req.Header.Add("Content-Encoding", string(compr))
+	req.Header.Set("Content-Type", contentTypeHeader(proto))
+	if proto == WriteProtoFullNameV1 {
+		// Compatibility mode for 1.0.
+		req.Header.Set(versionHeader, version1HeaderValue)
+	} else {
+		req.Header.Set(versionHeader, version20HeaderValue)
+	}
+
+	if attempt > 0 {
+		req.Header.Set("Retry-Attempt", strconv.Itoa(attempt))
+	}
+
+	resp, body, err := r.client.Do(ctx, req)
+	if err != nil {
+		// Errors from Client.Do are likely network errors, so recoverable.
+		return WriteResponseStats{}, retryableError{err, 0}
+	}
+
+	rs, err := parseWriteResponseStats(resp)
+	if err != nil {
+		r.opts.logger.Warn("parsing rw write statistics failed; partial or no stats", "err", err)
+	}
```

Comment:
should we only try to get write response stats/log a warning if the request type was v2?